vkcs/db_cluster.go

```diff
@@ -24,6 +24,69 @@ func flattenDatabaseClusterShard(inst dbClusterInstanceResp) map[string]interfac
 	return newShard
 }
 
+func flattenDatabaseClusterInstances(insts []dbClusterInstanceResp) []map[string]interface{} {
+	instances := make([]map[string]interface{}, len(insts))
+	for i, inst := range insts {
+		instances[i] = flattenDatabaseClusterInstance(inst)
+	}
+
+	return instances
+}
+
+func flattenDatabaseClusterInstance(inst dbClusterInstanceResp) map[string]interface{} {
+	instance := make(map[string]interface{})
+	instance["instance_id"] = inst.ID
+	instance["ip"] = inst.IP
+	instance["role"] = inst.Role
+
+	return instance
+}
+
+func expandDatabaseClusterShrinkOptions(v []interface{}) []string {
+	opts := make([]string, len(v))
+	for i, opt := range v {
+		opts[i] = opt.(string)
+	}
+	return opts
+}
+
+func databaseClusterDetermineShrinkedInstances(toDelete int, shrinkOptions []string, instances []dbClusterInstanceResp) ([]dbClusterShrinkOpts, error) {
+	ids := []dbClusterShrinkOpts{}
+	foundIDs := 0
+	if len(shrinkOptions) == 0 {
+		for _, instance := range instances {
+			if instance.Role != dbClusterInstanceRoleLeader {
+				ids = append(ids, dbClusterShrinkOpts{ID: instance.ID})
+				foundIDs++
+				if foundIDs == toDelete {
+					break
+				}
+			}
+		}
+		if foundIDs != toDelete {
+			return nil, fmt.Errorf("not enough instances to shrink cluster")
+		}
+	} else {
+		for _, instance := range instances {
+			needToRemain := false
+			for _, opt := range shrinkOptions {
+				if instance.ID == opt {
+					needToRemain = true
+				}
+			}
+			if !needToRemain {
+				ids = append(ids, dbClusterShrinkOpts{ID: instance.ID})
+				foundIDs++
+			}
+		}
+		if foundIDs != toDelete {
+			return nil, fmt.Errorf("invalid shrink options")
```

Comment:
explain more details in the message, e.g. "eigther wrong instance id, or wrong instance number specified" etc

Reply:
expanded error messages